vscode/src/qirGeneration.ts

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import * as vscode from "vscode";
+import { getCompilerWorker, log } from "qsharp-lang";
+import { isQsharpDocument } from "./common";
+
+let compilerWorkerScriptPath: string;
+
+export class QirGenerationError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = "QirGenerationError";
+  }
+}
+
+export async function getQirForActiveWindow(): Promise<string> {
+  let result = "";
+  const editor = vscode.window.activeTextEditor;
+  if (!editor || !isQsharpDocument(editor.document)) {
+    throw new QirGenerationError(
+      "The currently active window is not a Q# file"
+    );
+  }
+
+  // Check that the current target is base profile, and current doc has no errors.
+  const targetProfile = vscode.workspace
+    .getConfiguration("Q#")
+    .get<string>("targetProfile", "full");
+  if (targetProfile !== "base") {
+    throw new QirGenerationError(
+      "Submitting to Azure is only supported when targeting the QIR base profile. " +
+        "Please update the QIR target via the status bar selector or extension settings."
+    );
+  }
+
+  // Get the diagnostics for the current document.
+  const diagnostics = await vscode.languages.getDiagnostics(
+    editor.document.uri
+  );
+  if (diagnostics?.length > 0) {
+    throw new QirGenerationError(
+      "The current program contains errors that must be fixed before submitting to Azure"
+    );
+  }
+
+  const code = editor.document.getText();
+
+  // Create a temporary worker just to get the QIR, as it may loop/panic during codegen.
+  // Let it run for max 10 seconds, then terminate it if not complete.
```

Comment:
10 seconds seems like a rather aggresive timeout. Eventually I think we want a status popup that says we are compiling and allows cancellation. In the meantime I think something like a 30 second timeout might be more appropriate.

Reply:
You think? I actually had it at 5 seconds right until I pushed and decided to be more 'lenient'. With the speed we can run even complex shots at, what do you think might legitimately take up to 30 seconds to produce base profile code?

One we have telemetry I'll add a metric on compilation time (and timeouts). It would be an interesting value to observe.

Reply:
If the program is very heavy on classical computation and they are running on a lighter weight client, like VSCode on an iOS device via Safari + WASM, then it's possible they could bump into this. I agree that telemetry could make this an easier call, but for now I'd err on the side of more permissive of slower/longer computation.

Reply:
Done.